src/pages/index.js

```diff
@@ -0,0 +1,257 @@
+//   userInfo.setUserInfo(userData.name, userData.description);
+// });
+//
+
+// api.deleteCard()
+
+/*-----------------------------------------------------------------*/
+/*                             Validation                          */
+/*-----------------------------------------------------------------*/
+
+const validationSettings = {
+  inputSelector: ".modal__input",
+  submitButtonSelector: ".modal__button",
+  inactiveButtonClass: "modal__button_disabled",
+  inputErrorClass: "modal__input_type_error",
+  errorClass: "modal__error_visible",
+};
+
+const editProfileFormValidator = new FormValidator(
+  validationSettings,
+  editProfileForm
+);
+editProfileFormValidator.enableValidation();
+
+const addCardFormValidator = new FormValidator(
+  validationSettings,
+  addCardFormElement
+);
+addCardFormValidator.enableValidation();
+/*-----------------------------------------------------------------*/
+/*                             New Instances                       */
+/*-----------------------------------------------------------------*/
+
+//New Image Popup
+const newImagePopup = new PopupWithImage("#view-card-modal");
+newImagePopup.setEventListeners();
+function renderCard(data) {
+  const card = new Card(
+    data,
+    "#card-template",
+    handleCardClick,
+    handleDeleteClick,
+    handleLikeClick
+  );
+  return card.getView();
+}
+
+function handleLikeClick(card) {
+  const action = card._isLiked
+    ? api.removeLike.bind(api)
+    : api.likeCard.bind(api);
+  action(card._id)
+    .then((res) => {
+      card.updateLikes(res.isLiked);
+    })
+    .catch((err) => {
+      console.error(err);
+    });
+}
+
+// need an if/else statement that checks whether the card is currently liked or not.
+//hint: isLiked is true then we want to tell the server to remove a like
+// else we tell the server to remove the like
+// when finished, (inside .then) update view of heart button
+
+//New Card Popup
+const newCardPopup = new PopupWithForm(
+  "#add-card-modal",
+  handleAddCardFormSubmit
+);
+newCardPopup.setEventListeners();
+addNewCardButton.addEventListener("click", () => {
+  addCardFormValidator.resetValidation();
+  newCardPopup.openModal();
+});
+
+//Profile Edit Popup
+const newProfileEdit = new PopupWithForm(
+  "#profile-edit-modal",
+  handleProfileEditSubmit
+);
+newProfileEdit.setEventListeners();
+
+//Avatar Edit Popup
+const newAvatarEdit = new PopupWithForm(
+  "#edit-avatar-modal",
+  handleAvatarSubmit
+);
+avatarEditButton.addEventListener("click", () => {
+  newAvatarEdit.openModal();
+});
+const editAvatarValidator = new FormValidator(
+  validationSettings,
+  editProfileForm
+);
+
+/*-----------------------------------------------------------------*/
+/*                   "Handle" Functions                            */
+/*-----------------------------------------------------------------*/
+
+function handleProfileEditSubmit(inputValues) {
+  newProfileEdit.setSaving(true);
```

Comment:
**Could be improved**

If it’s interesting for you here is how we can make a universal function for handling any submit. We can get rid of such duplicating as **loading effect**, **closing** and **catching errors**


```js

// You can make a universal function that accepts a request, popup instance and optional loading text  
function handleSubmit(request, popupInstance, loadingText = "Saving...") {

  // here we change the button text
  popupInstance.renderLoading(true, loadingText);
  request()
    .then(() => {
      // We need to close only in `then`
      popupInstance.close()
    })
      // we need to catch possible errors
      // console.error is used to handle errors if you don’t have any other ways for that
     .catch(console.error)
    // in `finally` we need to return the initial button text back in any case
    .finally(() => {
      popupInstance.renderLoading(false);
    });
}



```


You need to pass the **request** into the call of `handleSubmit` like that:



```js

// here is an example of the profile form handling
function handleProfileFormSubmit(inputValues) {
  // we create a function that returns a promise
  function makeRequest() {
    // `return` lets us use a promise chain `then, catch, finally` inside `handleSubmit`
    return api.editProfile(inputValues).then((userData) => {
      userInfo.setUserInfo(userData)
    });
  }
  // Here we call the function passing the request, popup instance and if we need some other loading text we can pass it as the 3rd argument
  handleSubmit(makeRequest, profilePopup);
}



```